fluture.js

```diff
@@ -479,46 +442,178 @@
+    if(arguments.length === 1) return unaryPartial(parallel$i, i);
+    return parallel$i(i, ms);
   };
 
   Future.do = function Future$do(f){
-    if(!isFunction(f)) error$invalidArgument('Future.do', 0, 'be a function', f);
+    if(!isFunction(f)) invalidArgument('Future.do', 0, 'be a function', f);
     return new FutureDo(f);
   };
 
-  Future.chainRej = createUnaryDispatcher('chainRej');
-  Future.mapRej = createUnaryDispatcher('mapRej');
-  Future.swap = createNullaryDispatcher('swap');
-  Future.fork = createBinaryDispatcher('fork');
-  Future.race = createUnaryDispatcher('race');
-  Future.or = createUnaryDispatcher('or');
-  Future.fold = createBinaryDispatcher('fold');
-  Future.hook = createInvertedBinaryDispatcher('hook');
-  Future.finally = createUnaryDispatcher('finally');
-  Future.value = createUnaryDispatcher('value');
-  Future.promise = createNullaryDispatcher('promise');
-  Future.cache = createNullaryDispatcher('cache');
-  Future.extractLeft = createNullaryDispatcher('extractLeft');
-  Future.extractRight = createNullaryDispatcher('extractRight');
+  function chainRej$chainer(chainer, m){
+    if(!isFuture(m)) invalidArgument('Future.chainRej', 1, 'be a Future', m);
+    return new FutureChainRej(chainer, m);
+  }
+
+  Future.chainRej = function chainRej(chainer, m){
+    if(!isFunction(chainer)) invalidArgument('Future.chainRej', 0, 'be a Function', chainer);
+    if(arguments.length === 1) return unaryPartial(chainRej$chainer, chainer);
+    return chainRej$chainer(chainer, m);
+  };
+
+  function mapRej$mapper(mapper, m){
+    if(!isFuture(m)) invalidArgument('Future.mapRej', 1, 'be a Future', m);
+    return new FutureMapRej(m, mapper);
+  }
+
+  Future.mapRej = function mapRej(mapper, m){
+    if(!isFunction(mapper)) invalidArgument('Future.mapRej', 0, 'be a Function', mapper);
+    if(arguments.length === 1) return unaryPartial(mapRej$mapper, mapper);
+    return mapRej$mapper(mapper, m);
+  };
+
+  function race$left(left, right){
+    if(!isFuture(right)) invalidArgument('Future.race', 1, 'be a Future', right);
+    return new FutureRace(left, right);
+  }
+
+  Future.race = function race(left, right){
+    if(!isFuture(left)) invalidArgument('Future.race', 0, 'be a Future', left);
+    if(arguments.length === 1) return unaryPartial(race$left, left);
+    return race$left(left, right);
+  };
+
+  function or$right(right, left){
+    if(!isFuture(left)) invalidArgument('Future.or', 1, 'be a Future', left);
+    return new FutureOr(left, right);
+  }
+
+  Future.or = function or(right, left){
+    if(!isFuture(right)) invalidArgument('Future.or', 0, 'be a Future', right);
+    if(arguments.length === 1) return unaryPartial(or$right, right);
+    return or$right(right, left);
+  };
+
+  function finally$left(left, right){
+    if(!isFuture(right)) invalidArgument('Future.finally', 1, 'be a Future', right);
+    return new FutureFinally(left, right);
```

Comment:
These arguments were also flipped compared to the previous version, which was a far less obvious breaking change.